src/core/compact_object.cc

```diff
@@ -1037,7 +1040,8 @@ size_t CompactObj::MallocUsed() const {
 
   if (taglen_ == JSON_TAG) {
     DCHECK(u_.json_obj.json_ptr != nullptr);
-    return zmalloc_size(u_.json_obj.json_ptr);
+    auto* ptr = u_.json_obj.json_ptr;
+    return zmalloc_usable_size(ptr) + ptr->capacity();
```

Comment:
This is also wrong. `ptr->capacity` does not return the actual memory consumed in bytes. So clearly we have a problem. The main issues are:

1. There doesn't seem to exist a function in `Json` that clearly returns the size in bytes used by the object.
2. We can't approximate this easily since `json elements are polymoprhic` and it does not use flat storage internally such that we can use `zmalloc_usable_size()` on its `internal pointer`. 

We could in practice iterate over all of its elements and write a visitor that aggregates the size of all of its elements but that would be extremely slow especially when `info memory` command is used because now we need to iterate over all the json keys in the data store -- obviously a `no no`

And we also don't really control `json` since it's an external library. 

So my idea is to wrap `class MiMemoryResource` and introduce a `MiJsonMemoryResource`. This allocator will use the `same heap` as the `CompactObject::memory_resource` with one difference. It will have a member which will keep track how much memory it allocated/deallocated. Once the json operation finishes, we will reset this member value (such that subsequent operations on different json keys get accurate memory usage) and add/subtract it from a `size` object which will introduce in `JsonWrapper`. That way, when we ask for memory we will return this field without calculating anything and the memory consumption will properly be tracked by the allocator eliminating the need for manual tracking on object that belongs to a library we don't own. 

WDYT? @chakaz @adiholden

Reply:
Generally, using a custom memory resource for JSON sounds like a good solution. Indeed it would just call the data heap so all data will be in the same place as it was before.

This however won't play very nicely with the way we compute object memory today, and in many places we want to know the size of a specific object. So I wonder, perhaps we want to have a memory_resource per JSON object? It has the cost of 2 pointers per object, so not sure if it's worth it..

Reply:
> know the size of a specific object. 
Yep and that why will introduce a single size_t per json object in CompactObj. So the overhead will only be for one element which will keep the size (which is fine because we need a way to track memory for json objects)